### PR TITLE
Node controller to not force delete pods

### PR DESCRIPTION
**Pull request by foxish**
Fixes https://github.com/kubernetes/kubernetes/issues/35145

- [x] e2e tests to test Petset, RC, Job.
- [x] Remove and cover other locations where we force-delete pods within the NodeController.

**Release note**:

``` release-note
Node controller no longer force-deletes pods from the api-server.

* For StatefulSet (previously PetSet), this change means creation of replacement pods is blocked until old pods are definitely not running (indicated either by the kubelet returning from partitioned state, or deletion of the Node object, or deletion of the instance in the cloud provider, or force deletion of the pod from the api-server). This has the desirable outcome of "fencing" to prevent "split brain" scenarios.
* For all other existing controllers except StatefulSet , this has no effect on the ability of the controller to replace pods because the controllers do not reuse pod names (they use generate-name).
* User-written controllers that reuse names of pod objects should evaluate this change.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35235)

### CODE REVIEW

**Comment by AlexMioMio** [comment]
@foxish why we abandon this feature? i think force delete pod may be useful in some situation

**Comment by foxish** [comment]
@AlexMioMio, please see the discussion here https://github.com/kubernetes/kubernetes/issues/35145 and https://github.com/kubernetes/kubernetes/pull/34160, and feel free to comment on it. This is still a WIP.

**Comment by smarterclayton** [comment]
We are not abandoning this feature, but we are making it the administrator's responsibility to decide to force delete.  Node controller does not know whether the machine is partitioned and coming back or not, so if the node controller deletes those pods it can cause split brain in the cluster for pet sets.

**Comment by smarterclayton** [comment]
With this change, if you want to delete a node (because it's no longer running), delete the node and the pods will be cleaned up.  If you want to delete a pod, run `kubectl delete pod NAME --grace-period=0`

**Comment by foxish** [comment]
Closing this and reopening as part of this PR already got merged in https://github.com/kubernetes/kubernetes/pull/35541

**Comment by foxish** [comment]
Waiting for https://github.com/kubernetes/kubernetes/pull/35581 to merge before rebase. The other parts are now ready for review.

**Comment by smarterclayton** [comment]
Almost net negative change - awesome

**Comment by foxish** [comment]
@k8s-bot gci gce e2e test this

**Comment by foxish** [comment]
@k8s-bot gci gke e2e test this

**Comment by foxish** [comment]
@k8s-bot cvm gke e2e test this

**Comment by erictune** [comment]
Release note should have more detail I think.

Maybe add something like:
```
Node controller no longer force-deletes pods from the api-server.


For StatefulSet (previously PetSet), this change means creation of replacement Pods is blocked until old pods are definitely no running (indicated by node returning from partitioned state, or deletion of the Node object, or the instance in the cloud provider, or force deletion of the pod).
This has the desirable outcome of "fencing" to prevent "split brain" scenarios.

For all other existing controllers except StatefulSet , this has no effect on the ability of the controller to replace pods because the controllers do not reuse pod names (they use generate-name).

User-written controllers that reuse names should evaluate this change.
```

File: test/e2e/network_partition.go [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+	"k8s.io/kubernetes/test/e2e/framework"
+	testutils "k8s.io/kubernetes/test/utils"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// getMaster populates the externalIP, internalIP and hostname fields of the master.
+// If any of these is unavailable, it is set to "".
+func getMaster(c clientset.Interface) Address {
```
Comment (gmarek) [inline test/e2e/network_partition.go]: All those helper functions look like they belong to the `e2e/framework`.

File: test/e2e/network_partition.go [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+			})
+
+			// TODO marekbiskup 2015-06-19 #10085
+			// This test has nothing to do with resizing nodes so it should be moved elsewhere.
+			// Two things are tested here:
+			// 1. pods from a uncontactable nodes are rescheduled
+			// 2. when a node joins the cluster, it can host new pods.
+			// Factor out the cases into two separate tests.
+			It("[replication controller] recreates pods scheduled on the unreachable node "+
+				"AND allows scheduling of pods on a node after it rejoins the cluster", func() {
```
Comment (gmarek) [inline test/e2e/network_partition.go]: Where do we check if the Pod gets recreated?

File: test/e2e/network_partition.go [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+}
+
+// Blocks outgoing network traffic on 'node'. Then verifies that 'podNameToDisappear',
+// that belongs to replication controller 'rcName', really disappeared.
+// Finally, it checks that the replication controller recreates the
+// pods on another node and that now the number of replicas is equal 'replicas'.
+// At the end (even in case of errors), the network traffic is brought back to normal.
+// This function executes commands on a node so it will work only for some
+// environments.
+func performTemporaryNetworkFailure(c clientset.Interface, ns, rcName string, replicas int32, podNameToDisappear string, node *api.Node) {
```
Comment (gmarek) [inline test/e2e/network_partition.go]: Bad names. This function does way more than just performing TempNetworkFailure (it checks stuff), and IIUC Pod 'podNameToDisappear' should not disappear anymore, right?

**Comment by gmarek** [comment]
NodeController changes LGTM.

File: test/e2e/network_partition.go (hunk 3 above)
Comment (foxish) [inline test/e2e/network_partition.go]: The grace period is set to 0 for that pod in the RC, which is why it does still disappear. If it were set to default, or any other value, it wouldn't.

Comment (gmarek) [inline test/e2e/network_partition.go]: Then I think we're missing test case, right?

Comment (foxish) [inline test/e2e/network_partition.go]: @gmarek Yes, I'll be adding test-cases for job, rc, and so on today. :)

**Comment by foxish** [comment]
@gmarek Moved everything except the more esoteric helper functions to util, addressed comments and added new tests for RC and Job. PTAL.

**Comment by foxish** [comment]
@k8s-bot unit test this

**Comment by foxish** [comment]
@k8s-bot unit test this

**Comment by erictune** [comment]
lgtm

**Comment by smarterclayton** [comment]
Tests look fine to me, lgtm as well

**Comment by foxish** [comment]
Addressed comments by @erictune, will reapply LGTM after tests pass.

**Comment by foxish** [comment]
@k8s-bot node e2e test this

**Comment by SydOps** [comment]
I understand the request came from #35145 and #34160, but we get side effects because of this changes.

We implement kubernetes in aws autoscaling group. kubernetes nodes are keeping scaled up and down. If the terminated nodes keep NotReady status, a lot of daemonset pods are in pending status. 

If we start to disable this feature, and it is incompatible with old version, could we have the choices to decide, so we can have `force delete` as options when setup.

**Comment by smarterclayton** [comment]
That violates the safety guarantees of the cluster - if those nodes aren't coming back, you should delete the node.  Can you describe exactly why the nodes are in not ready state?  That seems like a flaw in the autoscaler or the cloud provider - if the instance is deleted in AWS, the cloud provider should be removing the instance from the API (the node should be deleted) which should clear the pods off the node, regardless of readiness.

**Comment by SydOps** [comment]
Thanks, @smarterclayton

We set up Kuberentes masters and nodes both with aws autoscaling group (ASG).  Version is 1.4.6. 

When scale down a node by ASG in aws, the node has been terminated, but the node stay in node list `kubectl get nodes` as `NotReady` status forever.

My colleague runs another kubernetes stack which is only v1.2, after the nodes are terminated by ASG as well in that environment, nodes disappear from node list in very short time.

I try to fix this issue, but failed. I think maybe the problem is related to the commit I mentioned above.

**Comment by foxish** [comment]
> When scale down a node by ASG in aws, the node has been terminated, but the node stay in node list kubectl get nodes as NotReady status forever.

I think that's the issue here. The nodes should be kept in sync with the underlying infrastructure always. The node object should be removed by either the cloud-provider specific code, or be deleted by an external loop. It shouldn't stick around indefinitely as "NotReady". 
The rationale for this change is explained also in https://kubernetes.io/docs/admin/node/#condition

In this case, I'd expect the cloud provider specific code to figure out that the instance is gone and remove the node object.
/cc @justinsb for AWS

**Comment by smarterclayton** [comment]
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/nodecontroller.go#L563
is what is supposed to be firing when we detect that the node no longer
exists in AWS

**Comment by foxish** [comment]
I confirmed that the node object does get deleted when the instance is deleted in case of GCE, GKE. So, the fault here is likely in the AWS cloud provider specific code.

**Comment by foxish** [comment]
I guess it is also possible that the instance is being stopped, but not deleted in this case. If that is the case, the cloud provider code would still find the instance, and the NC wouldn't delete it. Then the responsibility lies with the cluster admin to clean up such nodes.

**Comment by gmarek** [comment]
Maybe we should delete stopped Nodes as well?

**Comment by SydOps** [comment]
Thanks for the comments above. 

I am still not sure if this is related to my environment only. Because the PR is merged to version 1.5, my kubernetes version is 1.4.6 only.

**Comment by gmarek** [comment]
The only thing that was change here was handling Pods on not-ready Nodes. The code that is supposed to remove Nodes when they're not present in the cloud provider was untouched (i.e. Node should disappear when corresponding VM was deleted and cloud provider starts answering with "NotFound" when asked about it). We also moved the code that was responsible for deleting Pods from not existing Nodes, but that should have been no-op.

**Pull request merged**